Introduce QEMU for ARC EM & HS cores

abrodkin opened the pull request:
With QEMU port for ARC becoming more stable we're ready to open it and the first ever platform it's going to support will be Zephyr RTOS.

For starters we introduce bare minimal ARC EM & HS boards with just interrupts & UART but once other components get implemented and works stable enough we'll gradually add support for them in Zephyr's platform.

We're currently working to get all the legal approvals in place for distribution of QEMU off of Synopsys GitHub repo but to bridge the gap in time I suggest to grab QEMU from my repo here: https://github.com/abrodkin/qemu/releases/tag/arc-for-zephyr-0.1

This is based on a bit outdated upstream QEMU v4.2 but we'll rebase on top of v5.0 sometime soon.
To get QEMU binary for ARC one may just clone a repo mentioned above and then:
```
./configure --target-list=arc-softmmu && make
```

That will produce `qemu-system-arc` which needs to be added to a `PATH` env var, then it's possible to run sanity checker this way: ` ./scripts/sanitycheck --platform qemu_arcem` or ` ./scripts/sanitycheck --platform qemu_archs`

Commit:
ARC: Add support for basic QEMU platform for ARC EM & HS

Now when we're finally ready to open QEMU port for ARC
we introduce the first ever platform it supports and in fact does
that quite well - Zephyr RTOS.

For now we only offer support of basic EM & HS code execution,
built-in timers, interrupt controller and set of very simple
peripherals: DW UART & optionally MMIO Virtio devices.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>
Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

abrodkin commented:
@carlescufi I guess we're a bit late here for v2.3 right?

nashif commented:
you should call it qemu_arc, not plain qemu.

nashif commented:
> @carlescufi I guess we're a bit late here for v2.3 right?

no, you still have 1-2 days ;-)

nashif commented:
> you should call it qemu_arc, not plain qemu.

oh, I see you have different variants under qemu/

abrodkin commented:
@galak what should we do to add QEMU for ARC in the SDK?

galak commented:
> @galak what should we do to add QEMU for ARC in the SDK?

what version of QEMU is the baseline for ARC support?

abrodkin commented:
As of now this is something very close to v4.2. If it makes a difference I guess we may try to rebase on top of v5.0. Or which version is more suitable for Zephyr SDK?

galak commented:
> As of now this is something very close to v4.2. If it makes a difference I guess we may try to rebase on top of v5.0. Or which version is more suitable for Zephyr SDK?

the 0.11.x version is based on 4.2.  So if you want to submit a patch to Zephyr SDK for ARC support that should be doable.  Be aware we're looking to update to 5.0 soon.  So having that queued up as well would be great.

andrewboie commented:
> > As of now this is something very close to v4.2. If it makes a difference I guess we may try to rebase on top of v5.0. Or which version is more suitable for Zephyr SDK?
> 
> the 0.11.x version is based on 4.2. So if you want to submit a patch to Zephyr SDK for ARC support that should be doable. Be aware we're looking to update to 5.0 soon. So having that queued up as well would be great.

Moving to 5.0 would be great, how is testing for that coming along? Any showstoppers?

galak commented:
> Moving to 5.0 would be great, how is testing for that coming along? Any showstoppers?

I haven't tried it yet.  @stephanosio I think tried just ARM version.  Need to test all the qemu variants and see how it goes.

abrodkin commented:
@galak which repo/branch should I use for submitting ARC pathes for the SDK?

galak commented:
> @galak which repo/branch should I use for submitting ARC pathes for the SDK?

sdk-ng.  We don't have a git repo for qemu itself.  Just patches in sdk-ng.

See https://github.com/zephyrproject-rtos/sdk-ng/pull/200 for an example.

bwasim commented on boards/arc/qemu/doc/index.rst:
```diff
@@ -0,0 +1,100 @@
+
+Refer to the detailed overview about :ref:`application_debugging`.
+
+Networking
```
This heading should be removed I guess..

abrodkin replied:
Yet another copy-paste from ARM Cortex A53 docs. So frankly I have mo idea what's the right contents here, thus we're reviewing it here :)

Any good example or guide to look at?

bwasim replied:
All content looks okay.. I'm only pointing to remove the "Networking" heading because its not supported in your case. For reference, you can take a look at the `qemu_x86` readme file..

abrodkin replied:
Well as a matter of fact we may add networking with help of Virtio Ethernet controller - it's proven to work perfectly fine with Linux. But it looks like Virtio Ethernet controller driver is not yet implemented in the Zephyr, right? So then I'm dropping that header. Thanks for spotting this one!

bwasim commented on boards/arc/qemu/board.cmake:
```diff
@@ -0,0 +1,28 @@
+
+list(APPEND QEMU_FLAGS_${ARCH}
+  -M simhs
+  -m 128M
```
Invoking Qemu with `128M` but only using SRAM.. :-(

abrodkin replied:
Well that `-m 128M` is a left-over of what we used to use for all the rest including Linux use-case etc. Here I don't know which option to chose:
1. Decrease memory size in QEMU to match 256 KiB in the Zephyr or
2. Increase amount of memory claimed in Zephyr to 128 MiB
3. Something in between

What makes more sense in case of Zephyr?

bwasim replied:
Zephyr generally runs on low-end targets which have little Ram / Flash, and so you have no other option but to use them. However, in my opinion, this idea should not seep into platforms which have large amount of memory. In short, the bsp / board should be free to utilize the complete memory available on the board. 

There are cases where you don't want to use the complete memory - Mostly when you have multiple operating systems running on different cores of the same SoC. In these cases, you would have to "partition" the memory between both OSes.. But that's not applicable here.. 

So my opinion in this regard is to use the complete available memory.. May be others disagree ?

ruuddw replied:
Agree, the dts/linker files restrict already to a 'reasonable' size. Unless there's impact on memory/performance when configuring QEMU with larger memory, for flexbility it's fine to define the 128M here.

andrewboie replied:
128M is too much....can we drop this down to some reasonable value so that workstation ram usage doesn't blow up during highly parallel sanitycheck runs? On x86 we use 8M IIRC, there's no need for this to be so large

abrodkin replied:
@andrewboie right that makes perfect sense, I didn't consider multiple parallel sanitycheck runs on non-top notch full-blown servers :)

abrodkin replied:
@andrewboie done - down to 8 MiB in both .dts & QEMU command-line.

galak commented:
> > @galak which repo/branch should I use for submitting ARC pathes for the SDK?
> 
> sdk-ng. We don't have a git repo for qemu itself. Just patches in sdk-ng.
> 
> See [zephyrproject-rtos/sdk-ng#200](https://github.com/zephyrproject-rtos/sdk-ng/pull/200) for an example.

@abrodkin let me know what your plan is here so I know if I need to wait on a 0.11.3 SDK release.

abrodkin commented on boards/arc/qemu/qemu.dtsi:
```diff
@@ -0,0 +1,50 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		intc: arcv2-intc {
```
Agree, missed that. Will fix now.

abrodkin commented on boards/arc/qemu/qemu.dtsi:
```diff
@@ -0,0 +1,50 @@
+		interrupts = <24 1>;
+	};
+
+	chosen {
```
My idea was to keep as much as possible in the common .dtsi because what we have here is basically a virtual board with swappable CPU: we either use ARC EM-based CPU or ARC HS-based one. Note EM & HS are binary incompatible even though share the same (or very similar) ISA and so I don't have 2 completely separate boards for them. And that's what we have in case of nSIM as well BTW.

abrodkin commented on soc/arc/snps_qemu/soc.h:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @brief Board configuration macros for ARC QEMU
+ *
+ * This header file is used to specify and describe board-level
```
Copypasta detected, thanks :)

abrodkin commented:
@galak I'm already trying to integrate our huge patch (~2 MiB) in `sdk-ng/meta-zephyr-sdk/recipes-devtools/qemu` for 4.2-based case. So by the end of the day I hope to get it done.

Alternatively I noticed for Xilinx there's a separate target in https://github.com/zephyrproject-rtos/sdk-ng/tree/master/meta-zephyr-sdk/recipes-devtools/xilinx_qemu and so I may do a similar `arc_qemu` so that extarnal Git tree is used instead. But that's up to you to decide which way for me to go.

As for rebase on v5.0 I'd prefer to postpone it since we haven't done that rebase yet and so nobody knows how bad will be that can of worms :)

galak commented:
> @galak I'm already trying to integrate our huge patch (~2 MiB) in `sdk-ng/meta-zephyr-sdk/recipes-devtools/qemu` for 4.2-based case. So by the end of the day I hope to get it done.
> 
> Alternatively I noticed for Xilinx there's a separate target in https://github.com/zephyrproject-rtos/sdk-ng/tree/master/meta-zephyr-sdk/recipes-devtools/xilinx_qemu and so I may do a similar `arc_qemu` so that extarnal Git tree is used instead. But that's up to you to decide which way for me to go.
> 
> As for rebase on v5.0 I'd prefer to postpone it since we haven't done that rebase yet and so nobody knows how bad will be that can of worms :)

Gives me a rough sense which is what I was looking for.

abrodkin commented:
@galak just in case you're with that massive patch in `sdk-ng/meta-zephyr-sdk/recipes-devtools/qemu/files`.

abrodkin commented:
@galak see https://github.com/zephyrproject-rtos/sdk-ng/pull/208. It works quite well (~10 failures) with this PR :)

Let me know if something needs to be done more here to expedite things.

nashif commented:
There is still an issue with board naming, i build and run for qemu_arcem but I get:

```*** Booting Zephyr OS build v2.3.0-rc1-17-gdd14ea93c504  ***
Hello World! qemu
```

It should have qemu_arcem here.

nashif commented on boards/arc/qemu/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_QEMU
```
actually this is wrong, it should be BOARD_QEMU_ARCEM or something, lets not take this general name and make it ARC specific.  So start by calling the folder qemu_arcem and add everything to be for that board, then add the second configuration

abrodkin replied:
Renamed to `BOARD_QEMU_ARC` as it supports both ARC EM & ARC HS cores simultaneously.

stephanosio commented on boards/arc/qemu/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_QEMU
```
```suggestion
config BOARD_QEMU_ARC
```

abrodkin replied:
Fixed in the new re-spin.

stephanosio commented on boards/arc/qemu/Kconfig.defconfig:
```diff
@@ -0,0 +1,9 @@
+if BOARD_QEMU
+
+config BOARD
+	default "qemu"
```
`default qemu_arc`

abrodkin replied:
Fixed in the new re-spin. Thanks!

stephanosio commented on soc/arc/snps_qemu/Kconfig:
```diff
@@ -0,0 +1,15 @@
+	default SOC_QEMU_HS
+	depends on SOC_QEMU
+
+config SOC_QEMU_EM
```
```suggestion
config SOC_QEMU_ARC_EM
```

abrodkin replied:
Fixed in the new re-spin. Thanks!

stephanosio commented on soc/arc/snps_qemu/Kconfig:
```diff
@@ -0,0 +1,15 @@
+config SOC_QEMU_EM
+	bool "Synopsys ARC EM in nSIM"
+
+config SOC_QEMU_HS
```
```suggestion
config SOC_QEMU_ARC_HS
```

abrodkin replied:
Fixed in the new re-spin. Thanks!

stephanosio commented on soc/arc/snps_qemu/Kconfig.soc:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2020 Synopsys, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_QEMU
```
```suggestion
config SOC_QEMU_ARC
```

abrodkin replied:
Fixed in the new re-spin. Thanks!

stephanosio reviewed (dismissed):
Should not use `SOC_QEMU` and `BOARD_QEMU`. The symbol name should at least include `ARC`

abrodkin commented:
@nashif so this one won't get into v2.3, right?

abrodkin commented:
@stephanosio @nashif any better with recent changes?

nashif commented:
> @nashif so this one won't get into v2.3, right?

if we have the new SDK and this PR works fine once the SDK is enabled then I would ask for an exception from the TSC.

abrodkin commented:
@nashif QEMU for ARC is already in https://github.com/zephyrproject-rtos/sdk-ng/releases/tag/v0.11.3-beta-2. Pls try it out and let's see if it's OK for v2.3.

ruuddw commented on boards/arc/qemu/doc/index.rst:
```diff
@@ -0,0 +1,98 @@
+.. code-block:: console
+
+        *** Booting Zephyr OS build zephyr-v2.2.0-2486-g7dbfcf4bab57  ***
+        threadA: Hello World from qemu!
```
with the QEMU arch renamed, should this be "from qemu arc_hs" or something (didn't run the test myself yet)

abrodkin replied:
In fact it will be:
```
threadA: Hello World from qemu_arc!
threadB: Hello World from qemu_arc!
threadA: Hello World from qemu_arc!
threadB: Hello World from qemu_arc!
```

But thanks for the pointing this out :)
Will update the docs.

ruuddw commented on boards/arc/qemu/qemu.dtsi:
```diff
@@ -0,0 +1,50 @@
+/*
```
@galak: What's the recommendation here: dtsi file in the board directory, or in dts? I believe I've seen both.

ruuddw commented on soc/arc/snps_qemu/Kconfig:
```diff
@@ -0,0 +1,15 @@
+	depends on SOC_QEMU_ARC
+
+config SOC_QEMU_ARC_EM
+	bool "Synopsys ARC EM in nSIM"
```
typo / copy&paste mistake: nSIM -> QEMU (also below for HS)

abrodkin replied:
Copypasta :(
Fixed, thanks.

nashif commented:
now that we have ARC qemu in ci, job was restarted with the following issues:

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/73233/4/tests

abrodkin commented:
> now that we have ARC qemu in ci, job was restarted with the following issues:
> https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/73233/4/tests

With 6 failures IMHO that's quite a good result.

Moreover these are 2 assertions in `tests/kernel/tickless/tickless_concept/kernel.tickless.concept` and 4 failures in `tests/kernel/timer/timer_api/kernel.timer.tickless` which sorta hints at tickless timer support in QEMU for ARC. Hopefully this might be easily fixed sometime soon :)

abrodkin commented:
@nashif I cannot find a reference to Kconfig check results, any hints on what went wrong there (I'm worried because it says it's a **required** check above).

stephanosio commented:
> @nashif I cannot find a reference to Kconfig check results, any hints on what went wrong there (I'm worried because it says it's a **required** check above).

@abordkin https://github.com/zephyrproject-rtos/zephyr/pull/25100#issuecomment-625410013

nashif commented:
CONFIG_QEMU_TICKLESS_WORKAROUND was removed recently, you probably still have it in the PR

nashif commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -50,7 +50,8 @@
 
-#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL))
+#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL) &&		\
+		  !IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND))
```
this was removed due to icount introduction.

abrodkin replied:
@nashif sure I will remove entire commit `drivers: arcv2_timer0: Add work-around for QEMU` (https://github.com/zephyrproject-rtos/zephyr/pull/25100/commits/dd6b3a69919f3dc5661b9b74025bd42c0ebb1c69) then.

But the problem is we don't yet support `icount` (it's todo list but no ETAs yet for it) :(

andrewboie replied:
Be advised that lack of icount support does cause problems. We have quite a few timing sensitive tests that sporadically fall over if the host system is heavily loaded and QEMU keeps trying to sync the emulated clock to host time

nashif replied:
> @nashif sure I will remove entire commit `drivers: arcv2_timer0: Add work-around for QEMU` ([dd6b3a6](https://github.com/zephyrproject-rtos/zephyr/commit/dd6b3a69919f3dc5661b9b74025bd42c0ebb1c69)) then.
> 
> But the problem is we don't yet support `icount` (it's todo list but no ETAs yet for it) :(

if you need that workaround to pass the tests given you do not have icount I would suggest to reintroduce the workaround for ARC while we wait for icount.

abrodkin replied:
@nashif for starters I guess for ARC QEMU we may just disable `TICKLESS` config option until `icount` is implemented. What do you think?

abrodkin replied:
@nashif with rebase on top of the current master, removal of ARC Timer fix-up for QEMU & exclusion of `qemu_arc{em|hs}` in some tests we now have clean results, see https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/73352/

abrodkin commented:
@nashif looks like this is a "fix" for those failures along the lines of how this is "fixed" for some other QEMU platforms :)
```diff
diff --git a/tests/kernel/tickless/tickless_concept/testcase.yaml b/tests/kernel/tickless/tickless_concept/testcase.yaml
index 8bdfa1c0c4..62d0494cf0 100644
--- a/tests/kernel/tickless/tickless_concept/testcase.yaml
+++ b/tests/kernel/tickless/tickless_concept/testcase.yaml
@@ -5,5 +5,5 @@ tests:
     # consistently when coverage is enabled. Disable until 14173 is fixed.
     # This test fails on qemu_cortex_m0 consistently. Disable until the
     # root cause is identified.
-    platform_exclude: qemu_x86_coverage qemu_cortex_m0
+    platform_exclude: qemu_x86_coverage qemu_cortex_m0 qemu_arcem qemu_archs
     tags: kernel
diff --git a/tests/kernel/timer/timer_api/testcase.yaml b/tests/kernel/timer/timer_api/testcase.yaml
index e67b83d07a..cb130c5bd1 100644
--- a/tests/kernel/timer/timer_api/testcase.yaml
+++ b/tests/kernel/timer/timer_api/testcase.yaml
@@ -1,9 +1,9 @@
 tests:
   kernel.timer:
     tags: kernel userspace
-    platform_exclude: qemu_x86_coverage qemu_cortex_m0
+    platform_exclude: qemu_x86_coverage qemu_cortex_m0 qemu_arcem qemu_archs
   kernel.timer.tickless:
     extra_args: CONF_FILE="prj_tickless.conf"
     arch_exclude: riscv32 nios2 posix
-    platform_exclude: qemu_x86_coverage qemu_cortex_m0
+    platform_exclude: qemu_x86_coverage qemu_cortex_m0 qemu_arcem qemu_archs
     tags: kernel userspace
```

abrodkin committed:
tests: Exclude qemu_arc{em|hs} in some

See https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/73233/4/tests

And so until "icount" is implemented in QEMU for ARC we exclude
"tickless_concept" & "timer_api" from sanitycheck tests.

Signed-off-by: Alexey Brodkin <abrodkin@synopsys.com>

galak reviewed (changes requested):
Can we rename qemu_archs to qemu_arc_hs & qemu_arcem qemu_arc_em?

abrodkin commented:
> 
> 
> Can we rename qemu_archs to qemu_arc_hs & qemu_arcem qemu_arc_em?

@galak sure we may do that, it's just how we historically were naming ARC cores in open source projects (w/o space in between "arc" & "core_type/version"), see for example our `-mcpu` list:
```
$ arc-elf32-gcc --target-help

ARC-specific assembler options:
  -mcpu=<cpu name>        (default: arc700), assemble for CPU <cpu name>, one of:
                          arc700, nps400, arcem, em, em4, em4_dmips, em4_fpus,
                          em4_fpuda, quarkse_em, archs, hs, hs34, hs38,
                          hs38_linux, arc600, arc600_norm, arc600_mul64,
                          arc600_mul32x16, arc601, arc601_norm, arc601_mul64,
                          arc601_mul32x16
```

So that was more to keep things in line with other usecases.

ruuddw reviewed (approved):
Great, now we'll have not only build test CI coverage for ARC on PRs, but also run/execute tests!

galak commented:
> @galak sure we may do that, it's just how we historically were naming ARC cores in open source projects (w/o space in between "arc" & "core_type/version"), see for example our `-mcpu` list:
> 
> ```
> $ arc-elf32-gcc --target-help
> 
> ARC-specific assembler options:
>   -mcpu=<cpu name>        (default: arc700), assemble for CPU <cpu name>, one of:
>                           arc700, nps400, arcem, em, em4, em4_dmips, em4_fpus,
>                           em4_fpuda, quarkse_em, archs, hs, hs34, hs38,
>                           hs38_linux, arc600, arc600_norm, arc600_mul64,
>                           arc600_mul32x16, arc601, arc601_norm, arc601_mul64,
>                           arc601_mul32x16
> ```
> 
> So that was more to keep things in line with other usecases.

Is was mostly that archs looks like arch's and that's a bit confusing ;).

abrodkin commented:
> 
> 
> > @galak sure we may do that, it's just how we historically were naming ARC cores in open source projects (w/o space in between "arc" & "core_type/version"), see for example our `-mcpu` list:
> > ```
> > $ arc-elf32-gcc --target-help
> > 
> > ARC-specific assembler options:
> >   -mcpu=<cpu name>        (default: arc700), assemble for CPU <cpu name>, one of:
> >                           arc700, nps400, arcem, em, em4, em4_dmips, em4_fpus,
> >                           em4_fpuda, quarkse_em, archs, hs, hs34, hs38,
> >                           hs38_linux, arc600, arc600_norm, arc600_mul64,
> >                           arc600_mul32x16, arc601, arc601_norm, arc601_mul64,
> >                           arc601_mul32x16
> > ```
> > 
> > 
> > So that was more to keep things in line with other usecases.
> 
> Is was mostly that archs looks like arch's and that's a bit confusing ;).

Also if you take a look at how we instantiate QEMU you'll see the same spelling (`boards/arc/qemu/board.cmake`):
```
if(${CONFIG_SOC_QEMU_ARC_EM})
set(QEMU_CPU_TYPE_${ARCH} arcem)
set(QEMU_FLAGS_${ARCH} -cpu arcem)
elseif(${CONFIG_SOC_QEMU_ARC_HS})
set(QEMU_CPU_TYPE_${ARCH} archs)
set(QEMU_FLAGS_${ARCH} -cpu archs)
endif()
```

That said if renaming is a requirement or very strong suggestion I'll do it in a blink of an eye, otherwise I'd prefer to keep it as it is now.

galak commented:
> Also if you take a look at how we instantiate QEMU you'll see the same spelling (`boards/arc/qemu/board.cmake`):
> 
> ```
> if(${CONFIG_SOC_QEMU_ARC_EM})
> set(QEMU_CPU_TYPE_${ARCH} arcem)
> set(QEMU_FLAGS_${ARCH} -cpu arcem)
> elseif(${CONFIG_SOC_QEMU_ARC_HS})
> set(QEMU_CPU_TYPE_${ARCH} archs)
> set(QEMU_FLAGS_${ARCH} -cpu archs)
> endif()
> ```
> 
> That said if renaming is a requirement or very strong suggestion I'll do it in a blink of an eye, otherwise I'd prefer to keep it as it is now.

I get that, but the qemu flags are kinda hidden from most people, where as we're going to see qemu_archs in a number of sample.yaml, etc files.

galak commented:
I guess I'd also ask to rename the board dir from boards/arc/qemu to boards/arc/qemu_arc to match the convention we have on pretty much every other architecture.

vonhust reviewed (approved):
LGTM. really nice to have qemu support.

abrodkin commented:
@galak are we all good now or we want to have @andyross review or maybe even somebody else?

carlescufi commented:
> @carlescufi I guess we're a bit late here for v2.3 right?

As the release manager, I will say yes. But we are currently voting.

stephanosio commented:
Ran kernel tests on `qemu_arc_em` and `qemu_arc_hs` with the [QEMU 5.0 patch](https://github.com/zephyrproject-rtos/sdk-ng/pull/200/commits/ed10bba5d1b5acd77d7b149ef21c1ccd7c8cc6b0), and they all passed.

ARC QEMU 5.0 test build is available here:
https://github.com/stephanosio/zephyr-qemu/actions/runs/125474699

abrodkin commented:
> Ran kernel tests on `qemu_arc_em` and `qemu_arc_hs` with the [QEMU 5.0 patch](https://github.com/zephyrproject-rtos/sdk-ng/pull/200/commits/ed10bba5d1b5acd77d7b149ef21c1ccd7c8cc6b0), and they all passed.
> 
> ARC QEMU 5.0 test build is available here:
> https://github.com/stephanosio/zephyr-qemu/actions/runs/125474699

Wow, that's pretty cool! I mean building self-contained QEMU distribution for all 3 major platforms (Linux, Mac & Windows) with GitHub actions!
